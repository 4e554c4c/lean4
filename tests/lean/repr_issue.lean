def foo {m} [Monad m] [MonadExcept String m] [MonadState (Array Nat) m] : m Nat :=
catch (do modify $ λ a : Array Nat, a.write' 0 33,
          throw "error")
      (λ _, do a ← get, pure $ a.read' 0)

def ex₁ : StateT (Array Nat) (ExceptT String id) Nat :=
foo

def ex₂ : ExceptT String (StateT (Array Nat) id) Nat :=
foo

#exit

-- The following examples were producing an element of Type `id (Except String Nat)`.
-- Type class resolution was failing to produce an instance for `HasRepr (id (Except String Nat))` because `id` is not transparent.
#eval run ex₁ (mkArray 10 1000)
#eval run ex₂ (mkArray 10 1000)
