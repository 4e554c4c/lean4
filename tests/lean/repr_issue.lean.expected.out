repr_issue.lean:12:0: warning: using 'exit' to interrupt Lean
